eventstore/storethread.go

```diff
@@ -0,0 +1,145 @@
+package eventstore
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	tservopts "github.com/textileio/go-textile-core/options"
+	"github.com/textileio/go-textile-core/thread"
+	tserv "github.com/textileio/go-textile-core/threadservice"
+	threadcbor "github.com/textileio/go-textile-threads/cbor"
+	"github.com/textileio/go-textile-threads/util"
+)
+
+const (
+	addRecordTimeout  = time.Second * 10
+	fetchEventTimeout = time.Second * 15
+)
+
+// SingleThreadAdapter connects a Store with a Threadservice
+type singleThreadAdapter struct {
+	ctx      context.Context
+	api      tserv.Threadservice
+	store    *Store
+	threadID thread.ID
+	ownLogID peer.ID
+
+	lock    sync.Mutex
+	started bool
+}
+
+// NewSingleThreadAdapter returns a new Adapter which maps
+// a Store with a single Thread
+func newSingleThreadAdapter(ctx context.Context, store *Store, threadID thread.ID) *singleThreadAdapter {
+	a := &singleThreadAdapter{
+		ctx:      ctx,
+		api:      store.Threadservice(),
+		threadID: threadID,
+		store:    store,
+	}
+
+	return a
+}
+
+// Start starts connection from Store to Threadservice, and viceversa
+func (a *singleThreadAdapter) Start() {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+	if a.started {
+		return
+	}
+	a.started = true
+	li, err := util.GetOrCreateOwnLog(a.api, a.threadID)
+	if err != nil {
+		log.Fatalf("error when getting/creating own log for thread %s: %v", a.threadID, err)
+	}
+	a.ownLogID = li.ID
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go a.threadToStore(&wg)
+	go a.storeToThread(&wg)
+	wg.Wait()
+}
+
+func (a *singleThreadAdapter) threadToStore(wg *sync.WaitGroup) {
+	sub := a.api.Subscribe(tservopts.ThreadID(a.threadID))
+	defer sub.Discard()
+	wg.Done()
+	for {
+		select {
+		case <-a.ctx.Done():
+			log.Infof("cancelling thread event dispatch to store for thread %s since ctx was cancelled", a.threadID)
+			return
+		case rec, ok := <-sub.Channel():
+			if !ok {
+				log.Info("ending dispatch of events to store since thread channel was closed") // ToDo: reconsider action
```

Comment:
Huh, the threadservice closed the subscription channel. 
Opinions if this is a reasonable thing to do?
Currently, it will end processing thread->store direction of data.

Reply:
Seems ok to me since this will only happen if the thread service closes.